ledger-core/network/nds/msgdelivery/ctl_retry_msg_worker.go

```diff
@@ -106,78 +121,71 @@ func (p *retryMsgWorker) sendHead(msg *msgShipment, repeatFn func(retries.RetryI
 func noRetryFn(retries.RetryID) {}
 
 func (p *retryMsgWorker) processOoB(msg *msgShipment) {
-	var repeatFn func(retries.RetryID)
-
 	if msg.isFireAndForget() {
-		repeatFn = noRetryFn
+		p.processMsg(msg, noRetryFn)
 	} else {
-		repeatFn = p.sender.stages.AddHeadForRetry
+		p.processMsg(msg, p.sender.stages.AddHeadForRetry)
 	}
+}
 
-	switch {
-	case p.marks.mark(msg.id):
-		p.sema.Lock()
-		go p.sendHead(msg, repeatFn)
-	case msg.canSendHead():
-		p.pushPostponed(msg, repeatFn)
+func (p *retryMsgWorker) processMsg(msg *msgShipment, repeatFn func(retries.RetryID)) bool {
+	if msg == nil {
+		return true
 	}
-}
 
-func (p *retryMsgWorker) processMsg(msg *msgShipment, repeatFn func(retries.RetryID)) {
+	captured, postponer := p.marks.mark(msg.id, p)
 	switch {
-	case msg == nil:
-		return
-	case !p.marks.mark(msg.id):
-		//
+	case !captured:
 	case !p.sema.TryLock():
 		p.marks.unmark(msg.id)
 	default:
 		go p.sendHead(msg, repeatFn)
-		return
+		return true
 	}
 
 	if msg.canSendHead() {
-		p.pushPostponed(msg, repeatFn)
+		postponer.pushPostponed(msg, repeatFn)
```

Comment:
Why we push message to a worker which busy now? It some balancing mechanism?

Reply:
to keep all retries for the same node in a single worker - this will improve responsiveness of other workers on rescanning of postponeds